Lift user defined attributes into inputs for certain cases (user defined types and tensors)

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103711
* __->__ #103386

(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @msaroufim @aakhundov

voznesenskym committed:
wip

wip

[ghstack-poisoned]

voznesenskym committed:
Update on "wip"


wip

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

jansel commented on torch/_dynamo/variables/base.py:
```diff
@@ -163,6 +163,12 @@ def as_specialized(self, tx):
         return self
 
+    def associate(self, real_value):
+        self.real_value = real_value
```
could we get this lazily through the source rather than adding it as a property?

voznesenskym replied:
Yes, we absolutely could, a la `get` in guards.

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -197,9 +198,45 @@ def __init__(
-            # TODO(jansel): add guard for alias relationship
-            return self.tx.output.side_effects[value]
-        return self._wrap(value).clone(**self.options())
+            side_effect_result = self.tx.output.side_effects[value]
+            # Note - we may end up in a situation where we invoke something like
+            # def fn(x, y)
+            # with fn(x, x)
+            # Prior to the addition of tracking to all input objects, we would handle this just fine by
+            # eagerly re-entering VB and rewrapping inputs, correctly creating graphargs and placeholders. However,
+            # with tracking on inputs, we do not produce deduping guard correctly, and duplicate inputs or aliased
+            # relationships may end up getting erased here - the fn(x, x) call above, with side effects, would
+            # look like a graph with a single input (not necessarily incorrectly so).
```
self-note: rewrite

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

wconstab commented on torch/_dynamo/variables/base.py:
```diff
@@ -185,9 +185,53 @@ def const_getattr(self, tx, name: str) -> Any:
+        try:
+            _input_associated_real_value = eval(self.source.name(), scope)
+        except Exception:
+            # TODO(voz): Check in on this and find owner of SuperSource.
```
I added SuperSource a while back
https://github.com/pytorch/pytorch/commit/8e2e648f846f21aae7a8eeadf6964d8a9467af1b
The motivation chain was 
 - supporting default arguments required valid source chains for the function that has the default arg
 - some functions were created off of SuperVariables
 - we need a valid source for a SuperVariable
 
 do you have more info about what's broken?

voznesenskym replied:
I am not actually sold on super source always producing valid code! There are cases where it made code that failed to eval (The requirement for sources in dynamo)

voznesenskym replied:
I am working a PR to add this invariant atm.

voznesenskym commented:
https://github.com/pytorch/pytorch/actions/runs/5273644260

voznesenskym committed:
Update on "[Do not review] [WIP] Simple, lazy __dict__ simulation for dynamo inputs"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -197,9 +198,48 @@ def __init__(
+            )
+        return vt
+
+    def _can_lift_attrs_to_inputs(self, vt):
```
I plan to expand this function to cover more and more cases, but this is a good start for what we need it for atm.

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym commented:
All bench models pass, inductor 1-3% faster (due to fixing setattr graph break)

jansel commented on test/dynamo/test_subgraphs.py:
```diff
@@ -514,8 +514,8 @@ def fn(x, c1, c2, c3):
         # checking here we don't create 2^n graphs
-        self.assertEqual(cnt.frame_count, 7)
-        self.assertEqual(cnt.op_count, 10)
+        self.assertEqual(cnt.frame_count, 12)
```
Why did the frame count increase here?

voznesenskym replied:
Because we pass in a/b/c as aliases of the same value and dedup them - so we take the hit of guards and recompile when the bool value changes for one of the a/b/c

jansel commented on torch/_dynamo/variables/base.py:
```diff
@@ -185,9 +185,55 @@ def const_getattr(self, tx, name: str) -> Any:
+
+        from .builder import VariableBuilder
+
+        real_value = getattr(_input_associated_real_value, name)
```
`getattr` can run arbirary user code if `__getattribute__` `__getattr__` is defined.

voznesenskym replied:
Yea, we should check if its defined and reject, good point.

jansel commented on torch/_dynamo/variables/base.py:
```diff
@@ -185,9 +185,55 @@ def const_getattr(self, tx, name: str) -> Any:
+            # Note - at a certain point we may want to handle
+            raise NotImplementedError()
+
+        return VariableBuilder(tx, AttrSource(self.source, name))(
+            real_value
+        ).add_options(self)
```
We need some sort of guard that the attribute exists in a future call.

voznesenskym replied:
Yeah, I think a hasattr guard would not be amiss! Good point.

We have that in `HASATTR`

jansel commented on torch/_dynamo/variables/base.py:
```diff
@@ -185,9 +185,55 @@ def const_getattr(self, tx, name: str) -> Any:
         """getattr(self, name) returning a python constant"""
         raise NotImplementedError()
 
+    def dynamic_getattr(self, tx, name):
```
I don't think this should go on the base class.  Applying this to every single type seems like it could have unintended consequences.

voznesenskym replied:
Hmm, it passes tests and our CI suite, so I figured it was safe. Open to feedback though - is there somewhere else you'd like to see it moved to? Perhaps Tensors only?

jansel commented on torch/_dynamo/variables/builder.py:
```diff
@@ -197,9 +198,48 @@ def __init__(
-            # TODO(jansel): add guard for alias relationship
-            return self.tx.output.side_effects[value]
-        return self._wrap(value).clone(**self.options())
+            side_effect_result = self.tx.output.side_effects[value]
+            # Note - we may end up in a situation where we invoke something like
+            # def fn(x, y)
+            # with fn(x, x)
+            # Prior to the addition of tracking to all relevant objects, we would handle this just fine by
+            # eagerly re-entering VB and rewrapping inputs, correctly creating graphargs and placeholders. However,
+            # with tracking on inputs, duplicate inputs or aliased relationships may end up getting erased here -
+            # In the the fn(x, x) example call above look like a graph with a single input.
+            # In order to ensure that we do not reuse fn(x, x) for fn(x, y), we create a duplicate input guard.
+
+            # Note - we may not have a source, that is fine, it just means we had an object that is safe to have
+            # leave unsourced - like a local list created and discharged entirely within a local scope.
+            if side_effect_result.source and side_effect_result.source != self.source:
+                ser_source_is_local = is_from_local_source(side_effect_result.source)
+                source_is_local = is_from_local_source(self.source)
+                # Note - both must be local, or global, or we will run afoul of a lack of merging in how we currently
+                # reconcile guards builder scopes in compile_check_fn. This technically means we miss a guard here,
+                # so maybe we should do this refactor before we land this...
+                # TODO(voz): Combine local and global guard builders.
+                if ser_source_is_local == source_is_local:
+                    # Note - this is a little agressive - these being duplicate input does not always matter.
+                    # However, this should always be a sound guard to add here.
+                    dup_guard = functools.partial(
+                        GuardBuilder.DUPLICATE_INPUT, source_b=side_effect_result.source
+                    )
+                    side_effect_result = side_effect_result.add_guards(
+                        self.make_guards(dup_guard)
+                    )
+            return side_effect_result
+        vt = self._wrap(value).clone(**self.options())
+        if self._can_lift_attrs_to_inputs(vt):
+            vt = self.tx.output.side_effects.track_object_existing(
+                self.source, value, vt
+            )
+        return vt
```
I am having a hard time following what this code is doing.

It seems like there is some aliasing checking.  This doesn't seem like the right place to do that, also why is it only done for local sources?

There is also code to enable mutation tracking, but this seems like the wrong place to do that too.  Everywhere else we enable it in a type specific way not conditional on the source.

voznesenskym replied:
> I am having a hard time following what this code is doing.

This code intercepts VTs as they are made, and if they are an input, and of a specific kind we care about, adds attribute tracking (to do the input lifting)

What can I do to make this code clearer?

> It seems like there is some aliasing checking.

This is under the existing `if value in self.tx.output.side_effects:` - this is where we handle cases for when we deduped a value already - we are just installing a guard for the alias relationship (The todo you had)

> There is also code to enable mutation tracking, but this seems like the wrong place to do that too. Everywhere else we enable it in a type specific way not conditional on the source.

I don't want it for globals, and I only want it for certain types.

Do you want me to expand it to everywhere? It should be very doable, but I wanted to scope it down to a smaller change first and then expand `_can_lift_attrs_to_inputs`

jansel commented on torch/_guards.py:
```diff
@@ -146,11 +147,15 @@ def __hash__(self):
         return hash((self.name, self.source, id(self.create_fn)))
 
     def sort_key(self):
+        if isinstance(self.create_fn, functools.partial):
+            fn = self.create_fn.func
+        else:
+            fn = self.create_fn
```
set `firstlineno` here.  The name `fn` is misleading.

voznesenskym replied:
sure thing

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym commented:
Planned changes

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym committed:
Update on "Lift user defined attributes into inputs for certain cases (user defined types and tensors)"


(1) Lazy (converts to dynamo variable on access only)
(2) Uses existing side effect/reconstruct tech
(3) not tensor opinionated

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim aakhundov

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge
